### PR TITLE
mqtt over websocket on different path

### PR DESCRIPTION
**Pull request by namgk**
Fixes this #605

### CODE REVIEW

**Comment by namgk** [comment]
I don't see we have any test for attachHttpServer yet, I'm creating a new one then.

**Comment by namgk** [comment]
added some unit tests, however, there are a few things:

- if no path is specified in attachHttpServer, and a mqtt client connect to 'ws:<host>/<any_path>', should the server ignore the <any_path> part and still serve such client as if it was connecting to just 'ws:<host>'?
- was it your intention to always put a mqtt server on port 1883 as default (the case of empty option provided to mosca server)? This doesn't make much sense in the case where you are attaching to an existing http server.

File: package.json [hunk 1]
```diff
@@ -62,9 +63,9 @@
     "sinon-chai": "~2.8.0",
     "supertest": "~1.2.0",
     "tmp": "0.0.24",
-    "browserify": "~13.0.0",
     "uglify-js": "^2.4.16",
-    "underscore": "^1.7.0"
+    "underscore": "^1.7.0",
+    "ws": "^1.0.1"
```
Comment (mcollina) [inline package.json]: should be a real dependency, not a `devDependency`.

Comment (mcollina) [inline package.json]: forget about it.

Comment (namgk) [inline package.json]: So what do you mean? Would it be ok now then?

Comment (mcollina) [inline package.json]: it's ok, my bad.

**Comment by mcollina** [comment]
> if no path is specified in attachHttpServer, and a mqtt client connect to 'ws://<host>/<any_path>', should the server ignore the <any_path> part and still serve such client as if it was connecting to just 'ws://<host>'?

Keep the same behaviour as it was before, like it's now in code.

> was it your intention to always put a mqtt server on port 1883 as default (the case of empty option provided to mosca server)? This doesn't make much sense in the case where you are attaching to an existing http server.

Check out https://github.com/mcollina/mosca/blob/master/examples/Server_With_All_Interfaces-Settings.js#L16-L19.

**Comment by namgk** [comment]
I know it is possible to have server with all interface.

My point is that when I'm attaching to an existing http server, I don't want the port 1883 to be occupied, that's the reason why I'm using mqtt-over-ws.

**Comment by mcollina** [comment]
You can just specify the interfaces you want to listen to, or none.

**Comment by namgk** [comment]
the problem is if I specify none, there will still be a mqtt server listens on port 1883.

**Comment by namgk** [comment]
oh do you mean I should do 

```
var mqttServer = mosca.Server({interface:[]})
```
?
What I'm using now is

```
var mqttServer = mosca.Server({})
```

**Comment by mcollina** [comment]
Yes, `'interfaces'`,  not `'interface'`.

**Comment by namgk** [comment]
it still occupies port 1883. The following test failed:

```javascript
it.only("should not occupy 1883 port while attached to http server", function(done) {
    server = http.createServer();
    mqttServ = new mosca.Server({interfaces:[]});
    mqttServ.attachHttpServer(server);

    server.listen(port, function(){
      expect(port).to.not.equal(1883)
      var client = mqtt.connect('mqtt://localhost:1883');
      client.subscribe(mqttTopic);
      client.on("message", function(topic, payload) {
        done('failed')
      });
      client.publish(mqttTopic, ping);
    });
  });
```

**Comment by mcollina** [comment]
Would you mind sending a fix?

**Comment by namgk** [comment]
sure, I just want to make sure your intention in that case. I'll send another commit.

**Comment by namgk** [comment]
fixed empty interfaces options, refactored the tests to reflect this.

**Comment by namgk** [comment]
hi Matteo, do you have a plan to merge? Thanks

**Comment by mcollina** [comment]
I'm traveling till the end of this week, I will have a look afterwards.

File: lib/options.js [hunk 2]
```diff
@@ -259,11 +259,11 @@ function validate(opts, validationOptions) {
...
-  if (opts.hasOwnProperty('interfaces')) {
-    if (opts.interfaces.length === 0) {
-      result.addError('no interfaces were defined');
-    }
-  }
+  // if (opts.hasOwnProperty('interfaces')) {
+  //   if (opts.interfaces.length === 0) {
+  //     result.addError('no interfaces were defined');
+  //   }
+  // }
```
Comment (mcollina) [inline lib/options.js]: why this has been commented? If this is not needed anymore, we should just remove it.

**Comment by namgk** [comment]
Hi Matteo, that code block was under the assumption that there should be minimum one interface in the option. However this doesn't hold anymore due to the case where you are attaching the mqtt stack on an existing http server.

I deleted that code block.

**Comment by mcollina** [comment]
perfect, I'll try to assemble a release asap.

**Pull request merged**